Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/person/ui/overview/PersonOverviewViewModel.kt

```diff
@@ -0,0 +1,132 @@
+
+    private fun mapPersons(persons: Set<PersonCertificates>, qrCodesMap: Map<String, Bitmap?>): List<CertificatesItem> =
+        mutableListOf<CertificatesItem>().apply {
+            addPendingCards(persons)
+            addCertificateCards(persons, qrCodesMap)
+        }
+
+    private fun MutableList<CertificatesItem>.addCertificateCards(
+        persons: Set<PersonCertificates>,
+        qrCodes: Map<String, Bitmap?>
+    ) {
+        persons.filterNotPending()
+            .forEachIndexed { index, person ->
+                val certificate = person.highestPriorityCertificate
+                add(
+                    PersonCertificateCard.Item(
+                        certificate = certificate,
+                        qrcodeBitmap = qrCodes[certificate.qrCode],
+                        color = PersonOverviewItemColor.colorFor(index),
+                        onClickAction = {
+                            events.postValue(
+                                OpenPersonDetailsFragment(person.personIdentifier.codeSHA256)
+                            )
+                        }
+                    )
+                )
+            }
+    }
+
+    private fun PersonCertificates.hasPendingTestCertificate(): Boolean {
+        val certificate = highestPriorityCertificate
+        return certificate is TestCertificate && certificate.isCertificateRetrievalPending
+    }
+
+    private fun MutableList<CertificatesItem>.addPendingCards(persons: Set<PersonCertificates>) {
+        persons.forEach {
+            val certificate = it.highestPriorityCertificate
+            if (certificate is TestCertificate && certificate.isCertificateRetrievalPending) {
+                add(
+                    CovidTestCertificatePendingCard.Item(
+                        certificate = certificate,
+                        onRetryAction = { refreshCertificate(certificate.certificateId) },
+                        onDeleteAction = { events.postValue(ShowDeleteDialog(certificate.certificateId)) }
+                    )
+                )
+            }
+        }
+    }
+
+    private val PersonCertificatesProvider.qrCodesFlow
+        get() = personCertificates
+            .transform { persons ->
+                emit(emptyMap()) // Initial state
+                persons.filterNotPending()
+                    .forEach {
+                        val qrCode = it.highestPriorityCertificate.qrCode
+                        qrCodes[qrCode] = generateQrCode(qrCode)
+                        emit(qrCodes)
+                    }
+            }
+
+    private fun Set<PersonCertificates>.filterNotPending() =
+        filter {
+            !it.hasPendingTestCertificate()
+        }.sortedByDescending {
+            it.isCwaUser
+        }
+
+    private suspend fun generateQrCode(qrCode: QrCodeString): Bitmap? = try {
+        qrCodeGenerator.createQrCode(qrCode, margin = 0)
+    } catch (e: Exception) {
+        Timber.d(e, "generateQrCode failed for $qrCode")
+        null
```

Comment:
Should we show an error for this?

Reply:
it is a list , there are many qrcodes that need to be generated, maybe we need a placeholder in case of error ?

Reply:
Maybe there are just no errors :grin: ? How could get an error here?

Reply:
I did not get any 😃